core/src/main/scala/kafka/server/KafkaApis.scala

```diff
@@ -2180,6 +2186,26 @@ class KafkaApis(val requestChannel: RequestChannel,
       true
   }
 
+  def handleElectPreferredReplicaLeader(request: RequestChannel.Request): Unit = {
+
+    val electionRequest = request.body[ElectPreferredLeadersRequest]
+    val partitions =
+      if (electionRequest.topicPartitions() == null) {
+        zkClient.getAllPartitions()
+      } else {
+        electionRequest.topicPartitions().asScala
+      }
+    def sendResponseCallback(result: Map[TopicPartition, ApiError]): Unit = {
+      sendResponseMaybeThrottle(request, requestThrottleMs =>
+        new ElectPreferredLeadersResponse(requestThrottleMs, result.asJava))
+    }
+    if (!authorize(request.session, Alter, Resource.ClusterResource)) {
+      sendResponseCallback(partitions.map(partition => partition -> new ApiError(Errors.CLUSTER_AUTHORIZATION_FAILED, null)).toMap)
```

Comment:
Good point, but then since there are no TopicPartition which we know the client is authorized for it's a bit awkward to return the error. I propose to use a dummy `TopicPartition` (with null name and partition of -1) to avoid having to accommodate this in the `ElectPreferredLeadersResponse`.

Reply:
Perhaps we can just return an empty topic/partition list when the client asks for ALL partitions w/o permission. We did sth similar to that in handleTopicMetadataRequest.